graspologic/embed/n2v.py

```diff
@@ -0,0 +1,402 @@
+# Copyright (c) Microsoft Corporation and contributors.
+# Licensed under the MIT License.
+
+import logging
+import math
+import random
+import time
+from typing import Any, List, Optional, Tuple
+
+import networkx as nx
+import numpy as np
+
+
+def node2vec_embed(
+    graph: nx.Graph,
+    num_walks: int = 10,
+    walk_length: int = 80,
+    return_hyperparameter: int = 1,
+    inout_hyperparameter: int = 1,
+    dimensions: int = 128,
+    window_size: int = 10,
+    workers: int = 8,
+    iterations: int = 1,
+    interpolate_walk_lengths_by_node_degree: bool = True,
+) -> Tuple[np.ndarray, np.ndarray]:
+    """
+    Generates a node2vec embedding from a given graph. Will follow the word2vec algorithm to create the embedding.
+
+    Parameters
+    ----------
+
+    graph: nx.Graph
+        A networkx graph. If the graph is unweighted, the weight of each edge will default to 1
+    num_walks : int
+        Number of walks per source. Default is 10.
+    walk_length: int
+        Length of walk per source. Default is 80.
+    return_hyperparameter : int
+        Return hyperparameter (p). Default is 1.
+    inout_hyperparameter : int
+        Inout hyperparameter (q). Default is 1.
+    dimensions : int
+        Dimensionality of the word vectors. Default is 128.
+    window_size : int
+        Maximum distance between the current and predicted word within a sentence. Default is 10.
+    workers : int
+        Use these many worker threads to train the model. Default is 8.
+    iterations : int
+        Number of epochs in stochastic gradient descent (SGD)
+    interpolate_walk_lengths_by_node_degree : bool
+        Use a dynamic walk length that corresponds to each nodes
+        degree. If the node is in the bottom 20 percentile, default to a walk length of 1. If it is in the top 10
+        percentile, use walk_length. If it is in the 20-80 percentiles, linearly interpolate between 1 and walk_length.
+        This will reduce lower degree nodes from biasing your resulting embedding. If a low degree node has the same
+        number of walks as a high degree node (which it will if this setting is not on), then the lower degree nodes
+        will take a smaller breadth of random walks when compared to the high degree nodes. This will result in your
+        lower degree walks dominating your higher degree nodes.
+
+    Returns
+    Tuple[np.ndarray, np.ndarray]
+        A tuple containing a matrix, with each row index corresponding to the embedding for each node. The tuple
+        also contains a vector containing the corresponding vertex labels for each row in the matrix.
+        The matrix and vector are positionally correlated.
+    """
```

Comment:
One comment that maybe should be addressed prior to merge - In this docstring I really think we should cite the node2vec paper so we are attributing things properly

Reply:
We should also cite this code's genesis - the original implementation is mentioned in a non-docstring and we should credit the original author fully

Reply:
Actually I hadn't even copied the non-docstring comment - I'm fixing it now, but my last comment wasn't even right